content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md

```diff
@@ -0,0 +1,115 @@
+
+For authorization, this guide assumes that the cluser uses RBAC. If another authorization mode grants permissions to `kube-controller-manager` and `cloud-controller-manager` components, please grant the needed access in a way that matches the mode.
+
+<!-- steps -->
+
+### Grant access to Migration Lease
+
+The default permissions of the controller manager allow only accesses to their main Lease. In order for the migration to work, accesses to another Lease are required.
+
+You can grant `kube-controller-manager` full access to the leases API by modifying the `system::leader-locking-kube-controller-manager` role. This task guide assumes that the name of the migration lease is `cloud-provider-extraction-migration`.
+
+`kubectl patch -n kube-system role 'system::leader-locking-kube-controller-manager' -p '{"rules": [ {"apiGroups":[ "coordination.k8s.io"], "resources": ["leases"], "resourceNames": ["cloud-provider-extraction-migration"], "verbs": ["create", "list", "get", "update"] } ]}' --type=merge`
+
+Do the same to the `system::leader-locking-cloud-controller-manager` role.
+
+`kubectl patch -n kube-system role 'system::leader-locking-cloud-controller-manager' -p '{"rules": [ {"apiGroups":[ "coordination.k8s.io"], "resources": ["leases"], "resourceNames": ["cloud-provider-extraction-migration"], "verbs": ["create", "list", "get", "update"] } ]}' --type=merge`
+
+### Initial Leader Migration configuration
+
+Leader Migration requires a configuration file representing the state of controller-to-manager assignment. The following example configuration shows the assignment at this moment where `route`, `service`, and `cloud-node-lifecycle` controllers are started within `kube-controller-manager`.
+
+```yaml
+kind: LeaderMigrationConfiguration
+apiVersion: controllermanager.config.k8s.io/v1alpha1
+leaderName: cloud-provider-extraction-migration
+resourceLock: leases
+controllerLeaders:
+  - name: route
+    component: kube-controller-manager
+  - name: service
+    component: kube-controller-manager
+  - name: cloud-node-lifecycle
+    component: kube-controller-manager
+
+```
+
+On each control plane node, save the content to `/etc/leadermigration.conf`, and update the manifest of `kube-controller-manager` so that the file is mounted inside the container at the same location. Also, update the same manifest to add the following arguments:
+
+- `--feature-gates=ControllerManagerLeaderMigration=true` to enable Leader Migration which is an alpha feature
+- `--enable-leader-migration` to enable Leader Migration on the controller manager
+- `--leader-migration-config=/etc/leadermigration.conf` to set configuration file
+
+Restart `kube-controller-manager` on each node. At this moment, `kube-controller-manager` has leader migration enabled and is ready for the migration.
```

Comment:
Ideally this is done as a migration from version N (eg v1.21) to version N + 1 (eg v1.22). This should then be always be the KCM configuration at version N. This is important if you end up needing to rollback from N+1 to N. The migration work both upgrading and downgrading assuming it is configured this way.

Reply:
Worth mentioning the KCM should *NOT* be configured this way at version N+1.

Reply:
Will need to add "upgrade from N to N + 1" as an assumption to prerequisites and mention that N to N w/ CCM works too.

Reply:
"KCM should NOT be configured this way at version N+1" was mentioned below.

Reply:
Any reason why it is a bad idea to migrate from N w/ KCM to N w/ KCM+CCM?

Reply:
Depends on your tooling. It can be easier to track and automate things like upgrade/downgrade if its tied to a release. For example if you migrate at version N then need to downgrade for unrelated reasons, do you end up reverting to KCM only while not using the controller migration lock?

Reply:
Stated the reason to be "so that configuration can be aligned to releases"